api/versions.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2017 The WPT Dashboard Project. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+
+	"github.com/web-platform-tests/wpt.fyi/shared"
+	"google.golang.org/appengine/datastore"
+)
+
+// apiVersionsHandler is responsible for emitting just the browser versions for the test runs.
+func apiVersionsHandler(w http.ResponseWriter, r *http.Request) {
+	product, err := shared.ParseProductParam(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	} else if product == nil {
+		http.Error(w, fmt.Sprintf("Invalid product param: %s", r.URL.Query().Get("product")), http.StatusBadRequest)
+		return
+	}
+
+	ctx := shared.NewAppEngineContext(r)
+	query := datastore.NewQuery("TestRun").Filter("BrowserName =", product.BrowserName)
+	queries := []*datastore.Query{query}
+	if product.BrowserVersion != "" {
+		queries = make([]*datastore.Query, 2)
+		queries[0] = query.Filter("BrowserVersion =", product.BrowserVersion).Limit(1)
+		queries[1] = shared.VersionPrefix(query, "BrowserVersion", product.BrowserVersion, false).
```

Comment:
IIUC, `len(queries) == 1` before this line runs. Should this be `queries = append(queries, [...])`?

Reply:
You're right and wrong. I don't want the original solo query in the case that a version prefix is needed; I want "exactly that version" and "stuff starting with that version" (which I can't query for both of at once, because, Golang + datastore reasons).

Otherwise the query as built so far is the only one. I'll, uh, make it an else branch for readability.

Reply:
I guess what I don't understand is why isn't it true that "foo" starts with "foo" (because, it does, right?) If it's just something weird about Datastore queries though, then I'm happy to let the tests prove that this is necessary and sufficient.

Reply:
Because chrome-7 shouldn't match chrome-70, we cheat by using a prefix of `7.`
This leaves exactly one edge-case; a full string match.